File: scripts/lib/reference-utils.mjs
import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const booksPath = path.join(__dirname, 'canonical-books.json');
const books = JSON.parse(await readFile(booksPath, 'utf8'));

const aliasMap = new Map();
for (const book of books) {
  for (const alias of book.aliases) {
    aliasMap.set(normalizeBookToken(alias), book);
  }
}

export function getCanonicalBooks() {
  return books;
}

export function lookupBook(rawName) {
  const key = normalizeBookToken(rawName);
  return aliasMap.get(key) || null;
}

export function parseColonReference(rawReference) {
  const match = rawReference.trim().match(/^(?<book>.+?):(?<chapter>\d+)(?::(?<verses>\d+(?:-\d+)?))?$/);
  if (!match?.groups) return null;
  return buildReference(match.groups.book, match.groups.chapter, match.groups.verses);
}

export function parseSpaceUnderscoreReference(rawReference) {
  const match = rawReference.trim().match(/^(?<book>.+?)\s+(?<chapter>\d+)(?:_(?<verses>\d+(?:-\d+)?))?$/);
  if (!match?.groups) return null;
  return buildReference(match.groups.book, match.groups.chapter, match.groups.verses);
}

export function toReferenceKey(reference) {
  const base = `${reference.bookId}.${reference.chapter}`;
  if (!reference.verseStart) return base;
  if (reference.verseEnd && reference.verseEnd !== reference.verseStart) {
    return `${base}.${reference.verseStart}-${reference.verseEnd}`;
  }
  return `${base}.${reference.verseStart}`;
}

export function toChapterKey(reference) {
  return `${reference.bookId}.${reference.chapter}`;
}

export function toReferenceLabel(reference) {
  const chapterLabel = `${reference.bookName} ${reference.chapter}`;
  if (!reference.verseStart) return chapterLabel;
  if (reference.verseEnd && reference.verseEnd !== reference.verseStart) {
    return `${chapterLabel}:${reference.verseStart}-${reference.verseEnd}`;
  }
  return `${chapterLabel}:${reference.verseStart}`;
}

export function expandVerseRange(reference) {
  if (!reference.verseStart) return [];
  const end = reference.verseEnd || reference.verseStart;
  const results = [];
  for (let verse = reference.verseStart; verse <= end; verse += 1) {
    const verseReference = { ...reference, verseStart: verse, verseEnd: verse };
    results.push({
      ...verseReference,
      key: toReferenceKey(verseReference),
      label: toReferenceLabel(verseReference),
    });
  }
  return results;
}

export function normalizeBookToken(value) {
  return value
    .replace(/[.]/g, '')
    .replace(/\s+/g, ' ')
    .trim()
    .toLowerCase();
}

function buildReference(bookName, chapterText, versesText) {
  const book = lookupBook(bookName);
  if (!book) return null;
  const chapter = Number.parseInt(chapterText, 10);
  let verseStart = null;
  let verseEnd = null;
  if (versesText) {
    const [startText, endText] = versesText.split('-');
    verseStart = Number.parseInt(startText, 10);
    verseEnd = endText ? Number.parseInt(endText, 10) : verseStart;
  }
  return {
    bookId: book.id,
    bookName: book.name,
    canonical: book.canonical,
    chapter,
    verseStart,
    verseEnd,
  };
}

File: scripts/lib/reference_utils.py
from __future__ import annotations

import json
import re
from pathlib import Path


BOOKS_PATH = Path(__file__).with_name("canonical-books.json")
BOOKS = json.loads(BOOKS_PATH.read_text())


def _normalize_book_token(value: str) -> str:
    return re.sub(r"\s+", " ", value.replace(".", "")).strip().lower()


ALIAS_MAP = {}
for book in BOOKS:
    for alias in book["aliases"]:
        ALIAS_MAP[_normalize_book_token(alias)] = book


def lookup_book(raw_name: str):
    return ALIAS_MAP.get(_normalize_book_token(raw_name))


def parse_space_underscore_reference(raw_reference: str):
    match = re.match(r"^(?P<book>.+?)\s+(?P<chapter>\d+)(?:_(?P<verses>\d+(?:-\d+)?))?$", raw_reference.strip())
    if not match:
        return None
    return _build_reference(match.group("book"), match.group("chapter"), match.group("verses"))


def parse_colon_reference(raw_reference: str):
    match = re.match(r"^(?P<book>.+?):(?P<chapter>\d+)(?::(?P<verses>\d+(?:-\d+)?))?$", raw_reference.strip())
    if not match:
        return None
    return _build_reference(match.group("book"), match.group("chapter"), match.group("verses"))


def to_reference_key(reference: dict) -> str:
    base = f"{reference['bookId']}.{reference['chapter']}"
    verse_start = reference.get("verseStart")
    verse_end = reference.get("verseEnd")
    if not verse_start:
        return base
    if verse_end and verse_end != verse_start:
        return f"{base}.{verse_start}-{verse_end}"
    return f"{base}.{verse_start}"


def to_reference_label(reference: dict) -> str:
    chapter_label = f"{reference['bookName']} {reference['chapter']}"
    verse_start = reference.get("verseStart")
    verse_end = reference.get("verseEnd")
    if not verse_start:
        return chapter_label
    if verse_end and verse_end != verse_start:
        return f"{chapter_label}:{verse_start}-{verse_end}"
    return f"{chapter_label}:{verse_start}"


def _build_reference(book_name: str, chapter_text: str, verses_text: str | None):
    book = lookup_book(book_name)
    if not book:
        return None
    chapter = int(chapter_text)
    verse_start = None
    verse_end = None
    if verses_text:
        parts = verses_text.split("-")
        verse_start = int(parts[0])
        verse_end = int(parts[1]) if len(parts) > 1 else verse_start
    return {
        "bookId": book["id"],
        "bookName": book["name"],
        "canonical": book["canonical"],
        "chapter": chapter,
        "verseStart": verse_start,
        "verseEnd": verse_end,
    }
